src/usbsettings.cpp

```diff
@@ -31,17 +31,80 @@
 
 #include "usbsettings.h"
 
-USBSettings::USBSettings(QObject *parent)
-    : QObject(parent),
-      m_qmmode(new MeeGo::QmUSBMode(this))
+#include "usb_moded-dbus.h"
+#include "usb_moded-modes.h"
+
+#include <QDBusConnection>
+#include <QDBusMessage>
+#include <QDBusReply>
+
+
+static QDBusMessage call_usb_moded(const QString &method, const QList<QVariant> &arguments=QList<QVariant>())
+{
+    QDBusMessage methodCall = QDBusMessage::createMethodCall(USB_MODE_SERVICE, USB_MODE_OBJECT,
+            USB_MODE_INTERFACE, method);
+    methodCall.setArguments(arguments);
+    return QDBusConnection::systemBus().call(methodCall);
+}
+
+static const struct {
+    const char *name;
+    enum USBSettings::Mode mode;
+} mode_mapping[] = {
+    /* Keep this in sync with usb_moded-modes.h */
+    { MODE_MASS_STORAGE, USBSettings::MassStorage },
+    { MODE_CHARGING, USBSettings::ChargingOnly },
+    { MODE_CHARGING_FALLBACK, USBSettings::ChargingOnly },
+    { MODE_UNDEFINED, USBSettings::Undefined },
+    { MODE_ASK, USBSettings::Ask },
+    { MODE_DEVELOPER, USBSettings::Developer },
+    { MODE_MTP, USBSettings::MTP },
+    { MODE_CHARGER, USBSettings::Charger },
+    { MODE_HOST, USBSettings::Host },
+    { MODE_CONNECTION_SHARING, USBSettings::ConnectionSharing },
+    { MODE_DIAG, USBSettings::Diag },
+    { MODE_ADB, USBSettings::Adb },
+};
+
+static enum USBSettings::Mode decodeMode(const QString &name)
+{
+    for (int i=0; i<sizeof(mode_mapping)/sizeof(mode_mapping[0]); i++) {
```

Comment:
Compiler complains about comparing signed & unsigned values. Same for line 82. (I did a similar comment before, but I guess it was lost in transl... update).